lib/usagereporter/teleport/aggregating/service.go

```diff
@@ -70,6 +71,37 @@ func prepareUserActivityReport(
 	return report, nil
 }
 
+// resourcePresences returns the backend key for a resource presence report with
+// a given UUID and start time, such that reports with an earlier start time
+// will appear earlier in lexicographic ordering.
+func resourceCountsReportKey(reportUUID uuid.UUID, startTime time.Time) []byte {
+	return backend.Key(resourceCountsReportsPrefix, startTime.Format(time.RFC3339), reportUUID.String())
+}
+
+func prepareResourceCountsReport(
+	clusterName, reporterHostID []byte,
+	startTime time.Time, records []*prehogv1.ResourceCountRecord,
+) (*prehogv1.ResourceCountReport, error) {
+	reportUUID := uuid.New()
+	report := &prehogv1.ResourceCountReport{
+		ReportUuid:     reportUUID[:],
+		ClusterName:    clusterName,
+		ReporterHostid: reporterHostID,
+		StartTime:      timestamppb.New(startTime),
+		Records:        records,
+	}
+
+	for proto.Size(report) > maxItemSize {
+		if len(report.Records) <= 1 {
+			return nil, trace.LimitExceeded("failed to marshal resource counts report within size limit (this is a bug)")
+		}
+
+		report.Records = report.Records[:len(report.Records)/2]
```

Comment:
It is copied from userActivities reporting intact where it throws some “excess” data until report will become small enough to be stored.

It doesn't have any meaning when storing resource counts per kind as report size is constant in this case.

But if we change it to storing anonymized resource names as @espadolini suggested, then this size restriction become valid again: https://github.com/gravitational/teleport/blob/96ed71b2f7199f9ddc12b3ace6dee505eada5ddf/lib/usagereporter/teleport/aggregating/service.go#L31-L34

Reply:
Probably, for resource activity reports, we can split each report by resource kind, then one report for one resource kind can contain up to around 4k of anonymized resource names. Will it be enough?

Reply:
If it's not enough we're gonna have to split each report as well, unfortunately.

As far as the anonymized resource names go, we should also truncate them like we seem to have agreed upon in the yet-to-be-merged cloud rfd 88 https://github.com/gravitational/cloud/pull/6352/files#diff-6859f430e4c996afca151a4a59ffb7ad78613b952cc128a0bea1d4823fb3fa24R185-R222

Reply:
Ok, `fixed64` is 8 bytes instead of 32 bytes of HMAC-SHA-256, so it is around 16k of resources per kind per report. Probably still need to split, given that RFD 88 suggests that 100k resources of all kinds should be expected…

Reply:
We have to handle multiple reports with potentially overlapping and potentially disjoint sets of resources for the same time period anyway, since different agents will heartbeat through different auth servers (hopefully without losing connectivity to the one auth they're connected to, which means that only that one auth server will see that the resource exists). Handling multiple disjoint reports from the same auth server is no different than that.